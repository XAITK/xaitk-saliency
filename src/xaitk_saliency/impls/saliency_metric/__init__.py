"""Module for all implementations of the SaliencyMetric interface"""
