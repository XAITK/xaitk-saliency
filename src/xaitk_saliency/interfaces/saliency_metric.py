"""
This module defines the `SaliencyMetric` abstract base class, an interface for computing
metrics for a single saliency map input. Implementations of `SaliencyMetric` should
define the specific metric computation in the `compute` method, which can be called directly or
via the `__call__` method.

Classes:
    SaliencyMetric: An interface outlining the computation of a given metric for
    saliency map analysis.

Dependencies:
    - numpy for numerical operations on saliency maps.
    - smqtk_core for configuration management and plugin compatibility.

Example usage:
    class SpecificMetric(SaliencyMetric):
        def compute(self, sal_map):
            # Define metric calculation logic here.
            pass

    metric = SpecificMetric()
    score = metric(sal_map)
"""

from __future__ import annotations

import abc

import numpy as np
from smqtk_core import Plugfigurable


class SaliencyMetric(Plugfigurable):
    """
    This interface outlines the computation of a given metric when provided with
    a single input saliency map.
    """

    @abc.abstractmethod
    def compute(
        self,
        sal_map: np.ndarray,
    ) -> float:
        """
        Given up to two saliency maps, and additional parameters, return some given metric about
        the saliency map(s).

        :param sal_map: An input saliency map.

        :return: Returns a single scalar value representing an implementation's computed metric.
                 Implementations should impart no side effects upon the input saliency map.
        """

    def __call__(
        self,
        sal_map: np.ndarray,
    ) -> float:
        """Calls compute() with the given input saliency map(s) and additional parameters.

        :param sal_map: An input saliency map.

        :return: Returns a single scalar value representing an implementation's computed metric.
                 Implementations should impart no side effects upon the input saliency map.
        """
        return self.compute(sal_map)

    @property
    def name(self) -> str:
        """
        Returns the name of the SaliencyMetric instance.

        This property provides a convenient way to retrieve the name of the
        class instance, which can be useful for logging, debugging, or display purposes.

        Returns:
            str: The name of the SaliencyMetric instance.
        """
        return self.__class__.__name__
