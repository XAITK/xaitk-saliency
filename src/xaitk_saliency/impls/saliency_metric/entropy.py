"""
Provides an implementation of the `SaliencyMetric` interface for computing
entropy over a saliency map using `scipy.stats.entropy`.

Classes:
    Entropy: Computes the entropy of a given saliency map.

Example:
    >> entropy_saliency_metric = Entropy(clip_range=(0, 1))
    >> result = entropy_saliency_metric(sal_map)  # entropy_saliency_metric.compute(sal_map)
"""

from __future__ import annotations

from typing import Any

import numpy as np
from scipy.stats import entropy
from typing_extensions import override

from xaitk_saliency.interfaces.saliency_metric import SaliencyMetric


class Entropy(SaliencyMetric):
    """
    Implementation of the `SaliencyMetric` interface to calculate entropy.

    Entropy is a statistical measure of randomness that quantifies the amount of
    information in a saliency map. High entropy indicates complex, detailed information,
    while low entropy suggests uniform or less informative regions.

    Attributes:
        clip_range (tuple[float, float] | None): Optional (min, max) range to clip
            saliency values before computing entropy.
    """

    def __init__(self, clip_range: tuple[float, float] | None = None) -> None:
        """
        Initialize the Entropy metric.

        Args:
            clip_range (tuple[float, float] | None): Optional (min, max) value range to
                clip the saliency map before computing entropy. The metric computation
                does a min-max normalization if `clip_range` is not provided. If input
                already contains a [-1,1] normalized saliency map, the following clip
                ranges can be used:
                - Full Saliency: clip_range = (-1, 1)
                - Positive Saliency: clip_range = (0, 1)
                - Negative Saliency: clip_range = (-1, 0)
        """
        self.clip_range = clip_range
        if self.clip_range is not None and len(self.clip_range) != 2:
            raise ValueError("Invalid clip range. 2-element tuple of floats")

    @override
    def compute(
        self,
        sal_map: np.ndarray,
    ) -> float:
        """
        Compute the entropy of a saliency map.

        Entropy is computed using `scipy.stats.entropy` on the clipped/normalized saliency map.
        If `clip_range` is provided, the metric computation only performs range clipping and does
        not perform normalization. However, if `clip_range` is not provided, the metric computation
        performs only a min-max normalization.

        Args:
            sal_map (np.ndarray): The input saliency map.

        Returns:
            float: The entropy value of the saliency map.

        Warnings:
            UserWarning: If `ref_sal_map` is provided, it is ignored for entropy computation.

        """
        if self.clip_range is None:
            # The provided 1e-10 is a small non-zero value to prevent division by zero.
            s = (sal_map - sal_map.min()) / max((sal_map.max() - sal_map.min()), 1e-10)
        else:
            s = np.clip(sal_map, self.clip_range[0], self.clip_range[1])
        return float(entropy(s.ravel(), base=2))

    @override
    def get_config(self) -> dict[str, Any]:
        """
        Generates a configuration dictionary for the Entropy metric instance.

        Returns:
            dict[str, Any]: Configuration data representing the sensor and scenario.
        """
        return {
            "clip_range": self.clip_range,
        }
