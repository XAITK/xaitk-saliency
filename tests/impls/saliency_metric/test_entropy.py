from __future__ import annotations

from contextlib import AbstractContextManager
from contextlib import nullcontext as does_not_raise

import numpy as np
import pytest
from smqtk_core.configuration import configuration_test_helper
from syrupy.assertion import SnapshotAssertion

from tests.test_utils import CustomFloatSnapshotExtension
from xaitk_saliency.impls.saliency_metric.entropy import Entropy

from .saliency_metric_utils import saliency_metric_assertions

rng = np.random.default_rng()


@pytest.fixture
def snapshot_custom(snapshot: SnapshotAssertion) -> SnapshotAssertion:
    return snapshot.use_extension(CustomFloatSnapshotExtension)


class TestEntropy:
    """This class contains the unit tests for the functionality of the Entropy metric impl."""

    @pytest.mark.parametrize(
        ("sal_map", "expectation"),
        [
            (
                rng.integers(0, 255, (256, 256), dtype=np.uint8),
                does_not_raise(),
            ),
        ],
        ids=[
            "Single random input saliency map",
        ],
    )
    def test_consistency(
        self,
        sal_map: np.ndarray,
        expectation: AbstractContextManager,
    ) -> None:
        """Test Entropy metric call with various random saliency map inputs and parameters."""
        with expectation:
            inst = Entropy(clip_range=None)

            # Test the compute method directly
            compute_snr = saliency_metric_assertions(
                computation=inst.compute,
                sal_map=sal_map,
            )

            # Test callable
            callable_snr = saliency_metric_assertions(
                computation=inst,
                sal_map=sal_map,
            )

            assert compute_snr == callable_snr

    @pytest.mark.parametrize(
        ("clip_range"),
        [(None), ((-1.0, 1.0)), ((0.0, 1.0)), ((-1.0, 0.0))],
        ids=[
            "None",
            "Full Saliency Entropy",
            "Positive saliency entropy",
            "Negative saliency entropy",
        ],
    )
    def test_regression(
        self,
        clip_range: tuple[float, float] | None,
        snapshot_custom: SnapshotAssertion,
    ) -> None:
        """Regression testing results to detect API changes."""
        local_rng = np.random.default_rng(seed=12345)
        random_sal_map = local_rng.gumbel(loc=0, scale=0.1, size=(64, 64))
        random_sal_map /= random_sal_map.sum()
        inst = Entropy(clip_range=clip_range)
        entropy_metric_value = inst.compute(random_sal_map)
        snapshot_custom.assert_match(entropy_metric_value)

    @pytest.mark.parametrize(
        ("clip_range"),
        [(None), ((-1.0, 1.0))],
        ids=[
            "None",
            "Valid range tuple",
        ],
    )
    def test_configuration(self, clip_range: tuple[float, float] | None) -> None:
        """Ensure get_config is correct."""
        inst = Entropy(clip_range=clip_range)

        for inst_i in configuration_test_helper(inst):
            assert inst_i.clip_range == clip_range

    def test_classname(self) -> None:
        entropy_metric = Entropy(clip_range=None)
        assert entropy_metric.name == "Entropy"
