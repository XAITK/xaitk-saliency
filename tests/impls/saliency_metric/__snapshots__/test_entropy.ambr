# serializer version: 1
# name: TestEntropy.test_regression[Full Saliency Entropy]
  -inf
# ---
# name: TestEntropy.test_regression[Negative saliency entropy]
  10.160211041784207
# ---
# name: TestEntropy.test_regression[None]
  11.850504882120958
# ---
# name: TestEntropy.test_regression[Positive saliency entropy]
  10.87098660720397
# ---
